pkg/discovery/memcache/resolver.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package memcache
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type ClusterConfig struct {
+	version int
+	nodes   []Node
+}
+
+type Node struct {
+	dns  string
+	ip   string
+	port int
+}
+
+type Resolver interface {
+	Resolve(ctx context.Context, address string) (*ClusterConfig, error)
+}
+
+type memcachedAutoDiscovery struct {
+	dialTimeout time.Duration
+}
+
+func (s *memcachedAutoDiscovery) Resolve(ctx context.Context, address string) (config *ClusterConfig, err error) {
+	conn, err := net.DialTimeout("tcp", address, s.dialTimeout)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = conn.Close()
```

Comment:
If `err` is not nil then this nonetheless overrides the value of it, right? I suggest doing this here:
```
defer runutil.CloseWithErrCapture(&err, conn, "closing connection")
```

Reply:
thanks for the pro-tip